jupyter-mime.el

```diff
@@ -466,11 +478,11 @@ width and height of the image."
 ;;; `jupyter-insert' method additions
 
 (cl-defmethod jupyter-insert ((_mime (eql :text/html)) data
-                              &context ((functionp 'libxml-parse-html-region)
```

Comment:
Maybe we should instead prioritize the `jupyter-lang` context instead of removing this here so that whenever the `jupyter-lang` context is specified it means to override any other method and call that one first. The reason why you had to remove this context is because `(eql t)` takes precedence over `(jupyter-lang ...)`. If you change the 50 below

https://github.com/dzop/emacs-jupyter/blob/a775ac29bd5ab5a903a4e9c03116281a9b299aed/jupyter-client.el#L174

to 101 (100 is the priority of the `eql` specializer) and revert this change, the `R` method should be called first.

Reply:
I couldn't get this to work -- it looks like priority is ignored in `&context` specializers. Consider the following minimal example:
```
(cl-defgeneric foo (bar)
  (ignore))

(cl-defmethod foo (bar &context ((list t) (head t)))
  (message "hello"))

(cl-defmethod foo (bar &context (t (eql t)))
  (message "goodbye"))

(foo "")
```
This should print `"goodbye"` since `eql` has priority 100 while `head` has priority 80. However, it prints out `"hello"` when I test it -- `cl-generic.el` seems to sort the methods by the order they're defined and not by the priority.

Reply:
Yes that seems to be the case. I misunderstood the priority slot of a generalizer.

After spending some time trying to understand it better, the priority of the generalizers factor in when there are multiple generalizers that can be used for a particular specializer. The highest priority generalizer gets a chance at creating an approximation of an object for use with the specializer. In your example, the `head` specializer has a generalizer that takes `(list t)` and generalizes it into 
`(car (list t))` and thats what the `head` specializer looks at to determine if the method is valid.

But there can be multiple generalizers for a specializer, whichever one returns a non-nil value is used by the specializer and the generalizers are tried in the order of their priority.

Reply:
So it seems necessary to drop `&context` from the main `jupyter-insert` method for `text/html` output, unfortunately, because of the way priority works with `&context`. So I'm closing this thread. Please reopen if I'm wrong.